### PR TITLE
Game thread refactor

### PR DESCRIPTION
**Pull request by peppy**
Had to be done.

- New thread model, encapsulating Scheduler, PerformanceMonitor and Clock.
- Draw rate is now set to 4 * RefreshRate.
- Tidied up a lot of code.
- Alt-F4 exits on windows.
- Alt-Enter toggles exclusive fullscreen.
- Ctrl-F7 toggles frame limiters (draw and update).

### CODE REVIEW

File: osu.Framework/Platform/BasicGameWindow.cs [hunk 1]
```diff
@@ -64,6 +65,17 @@ private string GetVersionNumberSubstring(string version)
...
+        protected override void OnKeyDown(KeyboardKeyEventArgs e)
+        {
+            if (e.Key == Key.Enter)
+            {
+                WindowState = WindowState == WindowState.Fullscreen ? WindowState = WindowState.Normal : WindowState.Fullscreen;
+                return;
+            }
+
+            base.OnKeyDown(e);
+        }
```
Comment (peppy) [inline osu.Framework/Platform/BasicGameWindow.cs]: nice catch. must have lost a line somewhere along the way  >_>

File: osu.Framework/Platform/BasicGameHost.cs [hunk 2]
```diff
@@ -424,8 +343,8 @@ private void window_ClientSizeChanged(object sender, EventArgs e)
             Rectangle rect = Window.ClientRectangle;
             UpdateScheduler.Add(delegate
             {
-                //set base.Size here to avoid the override below, which would cause a recursive loop.
-                base.Size = new Vector2(rect.Width, rect.Height);
+            //set base.Size here to avoid the override below, which would cause a recursive loop.
+            base.Size = new Vector2(rect.Width, rect.Height);
```
Comment (peppy) [inline osu.Framework/Platform/BasicGameHost.cs]: the fuck

**Comment by SirCmpwn** [comment]
:+1:

File: osu.Framework/Platform/BasicGameHost.cs [hunk 3]
```diff
@@ -85,112 +61,72 @@ protected internal virtual Task SendMessage(IpcMessage message)
 
         public override bool IsVisible => true;
 
-        private static Thread updateThread;
```
Comment (Tom94) [inline osu.Framework/Platform/BasicGameHost.cs]: wait, we've been using static variables inside GameHost _already_ before? I'm surprised this has not affected the IPC tests. Why are those static again?

File: osu.Framework/Platform/BasicGameHost.cs [hunk 4]
```diff
@@ -204,43 +140,50 @@ protected int MaximumFramesPerSecond
...
         public static BasicGameHost GetInstanceIfExists() => instance;
 
-        protected virtual void OnActivated(object sender, EventArgs args)
-        {
-            UpdateScheduler.Add(delegate
-            {
-                IsActive = true;
-            });
-        }
+        protected virtual void OnActivated() => Schedule(() => setActive(true));
```
Comment (Tom94) [inline osu.Framework/Platform/BasicGameHost.cs]: I don't like that we are not seeing here anymore that we are scheduling to the update thread.

File: osu.Framework/Platform/BasicGameHost.cs [hunk 5]
```diff
@@ -341,80 +254,86 @@ protected virtual void DrawFrame()
...
+                {
+                }
             }
-            finally
+            else
             {
-                //if (!(error is OutOfMemoryException))
-                //    //we don't want to attempt a safe shutdown is memory is low; it may corrupt database files.
-                //    OnExiting();
+                while (!ExitRequested)
```
Comment (Tom94) [inline osu.Framework/Platform/BasicGameHost.cs]: Why are we not collecting performance statistics in this case?

File: osu.Framework/Platform/BasicGameHost.cs [hunk 6]
```diff
@@ -204,43 +140,50 @@ protected int MaximumFramesPerSecond
...
+                drawThread = new GameThread(DrawFrame, @"DrawThread")
+                {
+                    OnThreadStart = DrawInitialize,
+                },
+                updateThread = new GameThread(UpdateFrame, @"UpdateThread")
+                {
+                    OnThreadStart = UpdateInitialize,
+                    Monitor = { HandleGC = true }
+                },
+                inputThread = new InputThread(null, @"MainThread") //never gets started.
```
Comment (peppy) [inline osu.Framework/Platform/BasicGameHost.cs]: inputthread is main thread, therefore the gamethread isn't actually running as a thread (see the `InputThread` class used here).

File: osu.Framework/Platform/BasicGameHost.cs (hunk 5 above)
Comment (peppy) [inline osu.Framework/Platform/BasicGameHost.cs]: it's used in a headless context.

also WndProc is probably not even required in the above case (we don't run on the message processing thread any more due to the GameWindow not doing so), but I left it for now.

Comment (Tom94) [inline osu.Framework/Platform/BasicGameHost.cs]: This seems inconsistent, still. The update thread still collects performance statistics, for instance, when running headless. Hypothetically we may want to do performance regression test using performance counters in headless mode (although as you said, it'll be unlikely that this thread is even needed in headless mode).

Comment (peppy) [inline osu.Framework/Platform/BasicGameHost.cs]: the only thing it isn't collecting is "wndproc" which would not exist here.

**Pull request merged**